container/kvm/cache.go

```diff
@@ -0,0 +1,362 @@
+	i, err := maybeNewImage(md, pathfinder)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	dlURL, err := md.DownloadURL()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	req, err := http.NewRequest("GET", dlURL.String(), nil)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	return &updater{md: md, fileCache: i, req: req}, nil
+}
+
+// Sync updates the local cached images by reading the simplestreams
+// data and updating the locally cached images. It retrieves metadata
+// information from Oner and updates local cache(s) via Updater.
+func Sync(o Oner, u Updater) error {
+	md, err := o.One()
+	if err != nil {
+		return err
+	}
+	if u == nil {
+		u, err = newDefaultUpdater(md, paths.DataDir)
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+	err = u.Update()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	return nil
+}
+
+// Image represents a server image. In the KVM case a disk1.img, but could be
+// type: e.g. tar.gz.
+type Image struct {
+	Arch         string `yaml:"arch,omitempty"`
+	FilePath     string `yaml:"file_path,omitempty"`
+	MetadataPath string `yaml:"md_path,omitempty"`
+	FType        string `yaml:"ftype,omitempty"`
+	Series       string `yaml:"series,omitempty"`
+	ModTime      string `yaml:"mtime,omitempty"`
+	SHA256       string `yaml:"sha256,omitempty"`
+	Size         int64  `yaml:"size,omitempty"`
+	fileHandle   *os.File
+	mdHandle     *os.File
+}
+
+// Type returns the type of container this image is for. This is required by
+// for imagestorage.Metadata in the imagestorage.Storage.AddImage method call.
+func (i *Image) Type() string {
+	return "kmv"
+}
+
+// Close closes the file handles embedded in Image.
+func (i *Image) Close() error {
+	err := i.mdHandle.Close()
+	if err != nil {
+		err = errors.Trace(err)
+	}
+	err2 := i.fileHandle.Close()
+	if err2 != nil {
+		err = errors.Trace(err2)
+	}
+	return err
+}
+
+// write saves the stream to disk and updates the metadata file as well.
+func (i *Image) write(r io.Reader, md *imagedownloads.Metadata) error {
+	// First empty the existing file.
+	err := i.fileHandle.Truncate(0)
+	if err != nil {
+		return errors.Errorf("got %s when truncating %s",
+			err.Error(),
+			i.FilePath)
+	}
+	wrote, err := io.Copy(i.fileHandle, r)
+	if err != nil {
+		defer i.cleanup()
+		return errors.Trace(err)
+	}
+	if wrote != md.Size {
+		// This should likely be an error, with cleanup called prior to
+		// returning.
+		logger.Errorf("wrote %d, but expected to write %d", wrote, md.Size)
+	}
+	// We set ModTime to the time we got the update, now.
+	i.ModTime = time.Now().Format(http.TimeFormat)
+	// Size and hash are the only members that need updating.
+	i.Size = md.Size
+	i.SHA256 = md.SHA256
+	// TODO(ro) 2016-11-15 validate size and hash.
+	b, err := yaml.Marshal(i)
+	if err != nil {
+		defer i.cleanup()
+		return errors.Trace(err)
+	}
+	err = i.mdHandle.Truncate(0)
+	_, err = i.mdHandle.Write(b)
+	if err != nil {
+		defer i.cleanup()
```

Comment:
Just to clarify... i.cleanup() should only be called in defer if we encounter errors? We do not want to call it in just a defer at a method level?
I am wondering if these calls should be one call outside of if-err(s)...